Simplify simple template

joshka opened the pull request:
This will flow through to the async template too once the various issues here are resolved.

joshka commented on simple-generated/Cargo.toml:
```diff
@@ -6,4 +6,6 @@ license = "MIT"
 edition = "2021"
 
 [dependencies]
+crossterm = "0.28.1"
```
I think this is better to leave in based on seeing some of the comments in the issue related to this.

joshka commented on simple-generated/src/event.rs:
```diff
@@ -1,87 +1,116 @@
+/// thread. It sends a tick event at a regular interval. The tick rate is specified by the
+/// `tick_rate` parameter. If an error occurs, it sends an error event to the main thread and then
+/// stops running.
+fn event_thread(sender: mpsc::Sender<Event>, tick_rate: Duration) {
```
The order that things happen in this differ a little from the above implementation. This approach should give a tick event immediately, and then on the schedule. The above impl waits for the first timeout to send a tick event.

joshka commented on simple-generated/src/event.rs:
```diff
@@ -1,87 +1,116 @@
+            .map(|tick| tick.elapsed() >= tick_rate)
+            .unwrap_or(true)
+        {
+            sender.send(Event::Tick).expect("failed to send tick event");
```
This should ignore the error. Will fix.

joshka commented on simple-generated/src/event.rs:
```diff
@@ -1,87 +1,116 @@
+            tick_rate.saturating_sub(tick.elapsed())
+        });
+
+        match event::poll(timeout) {
+            Ok(false) => {
+                // no new events waiting
+                continue;
+            }
+            Ok(true) => {}
+            Err(e) => {
+                let _ = sender.send(Event::Error(e));
+                break;
+            }
+        }
+
+        match event::read() {
+            Err(err) => {
+                let _ = sender.send(Event::Error(err));
+                break;
+            }
```
This seems a bit verbose. I wonder if there's a more concise way to write this easily.

joshka commented on simple-generated/src/event.rs:
```diff
@@ -1,87 +1,116 @@
+            Ok(event) => match event {
+                CrosstermEvent::Key(e) if e.kind == KeyEventKind::Press => {
+                    // ignore key release / repeat events
+                    let _ = sender.send(Event::Key(e));
```
Send failures are expected and just indicate that the receiving side is closed. This is not an error that should cause a panic, it should gracefully just stop (perhaps the tick event not being able to be sent should just break out of the loop).

joshka commented on simple-generated/src/handler.rs:
Moved to main. This keeps the coherence of the modules high as the logic isn't spread across multiple files.

joshka commented on simple-generated/src/tui.rs:
-> ratatui::init

joshka commented on simple-generated/src/ui.rs:
moved to main

joshka commented on simple-generated/src/main.rs:
```diff
@@ -1,45 +1,111 @@
-    // Initialize the terminal user interface.
-    let backend = CrosstermBackend::new(io::stdout());
-    let terminal = Terminal::new(backend)?;
-    let events = EventHandler::new(250);
```
I pushed the event creation into the app, as main doesn't really need to know about it. It does need to know about terminal init / restore generally though as it's easier to handle restore nicely by return a result from the app method.

joshka reviewed (commented):
Some comments rationalizing the decisions here.

joshka commented:
I wonder whether this should be simplified further, to a point where it doesn't have the counter code. This way it would make a useful first step on the counter tutorial.

orhun commented on simple-generated/Cargo.toml:
```diff
@@ -6,4 +6,6 @@ license = "MIT"
 [dependencies]
+crossterm = "0.28.1"
 ratatui = "0.28.1"
+color-eyre = "0.6.3"
```
I personally don't like this—I believe we should leave the decision to use an error-handling library up to the user.

kdheepak replied:
The only reason I don't like color-eyre is because it's autocomplete messes with ratatui's styles. If it weren't for that, I'd always use color-eyre. I still think it is a good default for the template.

orhun commented on simple-generated/src/app.rs:
Having a separate `app.rs` will reduce the burden of splitting the state logic from `main.rs` later on. Most apps use a state to manage the layout and rendering, so I'd like to keep this in place but maybe rename it to `state.rs`.

joshka replied:
Moved back to app.rs in 5e8984d..ddb5fe3

orhun commented on simple-generated/src/main.rs:
```diff
@@ -1,45 +1,111 @@
-            Event::Mouse(_) => {}
-            Event::Resize(_, _) => {}
+
+fn main() -> Result<()> {
```
Why `main` is at the top of the file?

joshka replied:
;)
bottom
to
top
read
we
Because

(not the other way around)

kdheepak replied:
fwiw I dislike the top to bottom pattern. I tend to write code the way you would have to in C without explicit declarations, i.e. any function that needs to be used has to be defined above the function in which it is used. I also use this for structs, imports etc. And when I write it like this `main` always is at the end of the file. 

There are two reasons I dislike the top to bottom pattern, 1) it can the `main` somewhere in the middle of the file. 2) I don’t read code top to bottom sequentially in a file. I usually read it by tracing execution, i.e. I open the bottom of the file, scroll into `main`, use LSP to jump the definition of a function I’m interested in, jump back to main when I’m done, proceed.

joshka replied:
My reading order is find main and start reading in the same fashion as yours. It doesn't matter to me much where main starts. Cmd+F or Cmd+T works find. But I have an expectation that structs and enums are defined before their implementation. And I significantly dislike having to read functions and methods that are implemented above the code that calls them. When this happens, it feels like having to refer to a footnote in a book that's defined several pages previous. This interrupts the flow of my reading. It's not a linear top to bottom reading thing, but a cognitive burden thing. Functions and methods defined earlier in a file have a higher abstraction, more general, and have shorter names to reflect this (e.g. `main(), run(), init()`), while those lower in a file are more concrete and specific (`on_key_press()`, `increment_counter()`). I understand why C and other languages made this style (I learnt C++ as a kid in the 80s), but there are significant enough benefits of this top down approach that I think are worth always adhering to.

This practice generally is coupled with intention revealing / implementation hiding names for functions, which if done properly makes it easier to see the structure of an app.

I know in the past we've had similar discussions about:
```rust
fn foo() {
    // a
    some code related to a
    some code related to a
    some code related to a

    // b
    some code related to b
    some code related to b
    some code related to b

    // c
    some code related to c
    some code related to c
    some code related to c
}

/// vs
fn foo() {
    a();
    b();
    c();
}

fn a() {...}
fn b() {...}
fn c() {...}
```

This falls in the same bucket for me. Pulling out the code into methods puts a natural boundary around the possible interactions between each block of code. Arranging them in the calling order helps make that feel natural when reading too.
    
I'm definitely an opinionated datapoint of 1 here, but I've found that source code arranged this way seems to be the type that I find is significantly quicker to read diagonally. And it's because it's arranged like a good article is, the most important stuff is up front, the supporting stuff follows after. My metric really is based purely on how long it takes to understand the code when reading it later and how much of the code that I can skip reading rather than reading every single line of every single function.

kdheepak replied:
Minor rebuttal - I like `struct`s and `enum`s defined before they are used for the same reason I like functions to be defined before they are used. It's like the building blocks to form a program. Reading the file top to bottom builds cognitively the necessary information to understand what the program can do. 

I do agree that what you are saying has advantages too. At the end of the day it's probably two different styles? 

fwiw, I think writing code in "definition always before use" is simpler to follow. Also, for me another advantage is that it works in every language (python, C, C++, rust, nim, Julia etc) and I don't have to give it any more thought. I personally wouldn't choose a style that orders functions one way in C / C++ / nim and another way in Python, Rust, Julia.

kdheepak replied:
All that said, I'm indifferent to what approach we choose here! Most of my comments were related to reading `examples` code where everything is typically in one file, and in this case I think both ways are fine, since we are splitting content across files too.

orhun reviewed (changes requested):
Not super happy with this, I think having separate modules would work better.

orhun commented on simple-generated/src/event.rs:
```diff
@@ -1,87 +1,116 @@
+pub struct EventSource {
     /// Event receiver channel.
     receiver: mpsc::Receiver<Event>,
-    /// Event handler thread.
-    handler: thread::JoinHandle<()>,
```
Leaving this in would be the way of saying "hey you can actually access the event handler thread" like this.

joshka replied:
Can you give a concrete example of why a simple app would need to do that?

My goal with this template is to make it the sort of thing you'd reach for instead of cargo new. It should have the barest of things and should embody the YAGNI spirit. Another way of looking at this is to choose features where there is a demonstrated need, that need can be easily documented and supported by code that achieves this. Here it would be rare to need this.

kdheepak replied:
This is needed for killing the thread when you want to spawn vim, and need to spawn a new thread.

I agree about YAGNI spirit but I also would like to have it set up such that it extends well to future use cases. For example, when learning a new library or application, if I start with a simple tutorial example and find that it doesn't scale well or doesn't work with a feature I want to implement, I personally find that annoying. I do think it is quite common though.

joshka commented:
Where do the comments on this leave us? I suggest this is significantly better than what was there before and should be merged. Are there any comments that are hard blockers @orhun @kdheepak ?

orhun commented:
This template still leaves some beginner questions unanswered and might leave the user confused:

- **How do you send custom events if needed?** This is a common issue.
  - We should consider exposing `mpsc::Sender` somewhere in the app. Alternatively, we could show how to use `send()`.
  
- **What is `color_eyre`?** I came to use `ratatui`, but now there's another dependency I need to learn about!
  - We could either remove `color_eyre` or provide documentation explaining its purpose.
  
- **The `ui()` function tends to become very long after a while.** Perhaps we could move it to a `ui.rs` file? This might make it easier to locate and search for.
  
- **Why not use a constant for FPS?**

joshka commented:
> How do you send custom events if needed? This is a common issue.

In the simple template, you don't

> What is color_eyre? I came to use ratatui, but now there's another dependency I need to learn about!

We do document this, in the second counter tutorial, in https://ratatui.rs/recipes/apps/color-eyre/, in https://github.com/ratatui/ratatui/blob/main/examples/README.md#design-choices (documenting why we use this in all the examples). Additionally hovering over the color_eyre line shows a decent message that answers the question.

> The ui() function tends to become very long after a while. Perhaps we could move it to a ui.rs file? This might make it easier to locate and search for.

I have a strong (evidence based) opinion against pre-emptively organizing a project into horizontal layers. Having worked on many projects small and large, horizontal segmentation leads to poorer long term maintainability for a variety of reasons. Doing this would be pre-emptive. Put another way, if I was using the simple template personally, having the extra file would be an annoyance that I'd always want to revert. Reverting would be annoying. Refactoring in the other direction is fairly simple though - highlight code, extract to module, done.

> Why not use a constant for FPS?

6ba914b

joshka commented:
Moved the template under simple/template so that we can have a readme for the template and a generated readme for the template user.

Added documentation for color-eyre, fixed up the readme outline of the files, added github workflows and dependabot settings.

Removed cargo.lock as it will get out of sync pretty quickly and running cargo on the generated source will effectively update to the latest immediately.

orhun commented:
I'm not sure adding GitHub workflows to the templates. It might increase complexity, especially for new users who would need to understand what those 100 lines of opinionated YAML is doing, how to use GitHub Actions, etc. Additionally, it's not directly related to the context of TUIs as well. Personally, I’d find it annoying and would likely delete the `.github` folder before starting to work on the template. Thoughts? @kdheepak @joshka

joshka commented:
This is a template, stored on github to create an application. Generating github workflows to ensure that the project builds well is a useful thing to do. Anyone using GitHub should have a basic understanding of github workflows, which are well documented.

The "opinionated" parts of these are fairly insignificant and the sort that I wonder why you'd delete. They just check formatting, check clippy lints, check docs, and run any tests. These are basic hygiene things for any rust project.

Given that deleting these is a matter of deleting a single file, and that this generally makes any project better, I'm for leaving these in.

kdheepak commented on simple-generated/src/event.rs:
```diff
@@ -1,87 +1,116 @@
+    Error(std::io::Error),
 }
 
 /// Terminal event handler.
-#[allow(dead_code)]
+///
+/// This struct is responsible for listening to terminal events and sending them to the main thread.
+///
+/// It sends a tick event at a regular interval. The tick rate is specified by the `tick_rate`
+/// parameter. If an error occurs, it sends an error event to the main thread and then stops
+/// running.
 #[derive(Debug)]
-pub struct EventHandler {
-    /// Event sender channel.
-    sender: mpsc::Sender<Event>,
+pub struct EventSource {
```
Maybe `EventListener`? 

I actually like `EventHandler`, because it technically does handle getting the events and allows the user to retrieve them after it gets them.

I asked chatgpt for various names:

<img width="850" alt="image" src="https://github.com/user-attachments/assets/d9f2d17a-7333-4e8d-bf46-cf81c71da0df">

joshka replied:
Removed the entire module in 6a76940

kdheepak commented:
I made a couple of minor comments but it looks good to me overall.

orhun commented:
> The "opinionated" parts of these are fairly insignificant and the sort that I wonder why you'd delete. They just check formatting, check clippy lints, check docs, and run any tests. These are basic hygiene things for any rust project.

It just felt a bit too much for a simple project creation, that's all. But yes, I agree with your point that the user can just delete them, so it shouldn't be a big deal.

orhun reviewed (approved):
Slapping an approval here not to block this further, might come back to it later :)

joshka reviewed (commented):
Removed the counter behavior and the events module entirely as it was not a "simple" implementation and It's not really necessary for simple apps at all to worry about a tick event. This makes this a good basis for the simple tutorials.

joshka commented:
Let's KISS on this template, and move the more opinionated ideas about event handling, threads, etc. to another more advanced template that documents them well.

joshka commented:
Merging this to be able to rewrite the getting started docs.

orhun commented:
Sounds good, thanks for your work on this 👍🏼
